packages/react/src/components/DatePicker/next/DatePicker.js

```diff
@@ -0,0 +1,656 @@
+          inputFrom: startInputField.current,
+          inputTo: endInputField.current,
+        }),
+      ],
+      clickOpens: true,
+      nextArrow: rightArrowHTML,
+      prevArrow: leftArrowHTML,
+      onChange: savedOnChange,
+      onClose: savedOnClose,
+      onReady: onHook,
+      onMonthChange: onHook,
+      onYearChange: onHook,
+      onOpen: (...args) => {
+        onHook(...args);
+        savedOnOpen(...args);
+      },
+      onValueUpdate: onHook,
+    });
+
+    calendarRef.current = calendar;
+
+    function handleArrowDown(event) {
+      if (match(event, keys.ArrowDown)) {
+        const {
+          calendarContainer,
+          selectedDateElem: fpSelectedDateElem,
+          todayDateElem: fptodayDateElem,
+        } = calendar;
+        const selectedDateElem =
+          calendarContainer.querySelector('.selected') && fpSelectedDateElem;
+        const todayDateElem =
+          calendarContainer.querySelector('.today') && fptodayDateElem;
+        (
+          selectedDateElem ||
+          todayDateElem ||
+          calendarContainer.querySelector('.flatpickr-day[tabindex]') ||
+          calendarContainer
+        ).focus();
+      }
+    }
+
+    function handleOnChange() {
+      if (start.value !== '') {
+        return;
+      }
+
+      if (!calendar.selectedDates) {
+        return;
+      }
+
+      if (calendar.selectedDates.length === 0) {
+        return;
+      }
+
+      calendar.clear();
+      calendar.input.focus();
+    }
+
+    if (start) {
+      start.addEventListener('keydown', handleArrowDown);
+      start.addEventListener('change', handleOnChange);
+
+      // Flatpickr's calendar dialog is not rendered in a landmark causing an
+      // error with IBM Equal Access Accessibility Checker so we add an aria
+      // role to the container div.
+      calendar.calendarContainer.setAttribute('role', 'region');
+      // IBM EAAC requires an aria-label on a role='region'
+      calendar.calendarContainer.setAttribute(
+        'aria-label',
+        'calendar-container'
+      );
+    }
+
+    if (end) {
+      end.addEventListener('keydown', handleArrowDown);
+      end.addEventListener('change', handleOnChange);
+    }
+
+    //component did unmount equivalent
+    return () => {
+      // Note: if the `startInputField` ref is undefined then calendar will be
+      // of type: Array and `destroy` will not be defined
+      if (calendar && calendar.destroy) {
+        calendar.destroy();
+      }
+
+      if (start) {
+        start.removeEventListener('keydown', handleArrowDown);
+        start.removeEventListener('change', handleOnChange);
+      }
+
+      if (end) {
+        end.removeEventListener('keydown', handleArrowDown);
+        end.removeEventListener('change', handleOnChange);
+      }
+    };
+  }, [savedOnChange, savedOnClose, savedOnOpen]); //eslint-disable-line react-hooks/exhaustive-deps
+
+  // component did update equivalent
+  useEffect(() => {
```

Comment:
Totally get what you mean by what you were saying the other day 😭  This is such a tough one to refactor.

I feel like we have only a handful of options that we can realistically do:

- One effect per calendar setting

```jsx
useEffect(() => {
  if (calendarRef.current) {
    calendarRef.current.set({ dateFormat });
  }
}, [dateFormat]);

useEffect(() => {
  if (calendarRef.current) {
    calendarRef.current.set({ minDate });
  }
}, [minDate]);

// ...
```

- Pros
  - Simple approach, each setting gets an effect
  - Simple to extend as we just can add on using the existing pattern
- Cons
  - Cumbersome, we have so many fields 😭 
  - Awkward "first render" problem where we only want to run this code _after_ the first render


- We could create a custom way to track these changes

```jsx
function useTrackChanges(dependencies) {
  const [changes, setChanges] = useState([]);
  const [snapshot, updateSnapshot] = useState(dependencies);

  useEffect(() => {
    const changes = Object.keys(dependencies)
      .filter((key) => {
        return dependencies[key] !== snapshot[key];
      })
      .map((key) => {
        return {
          key,
          value: dependencies[key],
        };
      });

    if (changes.length > 0) {
      setChanges(changes);
    }

    updateSnapshot(dependencies);
  }, [dependencies]);

  return changes;
}

function DatePicker({ dateFormat, minDate }) {
  const changes = useTrackChanges({
    dateFormat,
    minDate,
    // ...
  });

  useEffect(() => {
    changes.forEach(change => {
      const { key, value } = change;
      calendarRef.current.set({ [key]: value });
    });
  }, [changes]);
}
```

- Pros
  - Simpler `useEffect` for adding changes to the calendar
- Cons
  - We're working outside of React's system (a little bit)
  - Since the object identity may not always be stable, we may be doing equality checks that we may not need (fixable with useMemo, however)

---

When you're looking at some of these options and what you have so far, are there any that stand out to you or others that you think we should consider? 👀